qiita_db/handlers/oauth2.py

```diff
@@ -0,0 +1,280 @@
+        """Create the access token for the client on redis
+
+        Parameters
+        ----------
+        token: str
+            Random token string for authorization
+        client_id : str
+            Client that requested the token
+        user : str, optional
+            If password grant type requested, the user requesting the key.
+        timeout : int
+            The timeout, in seconds, for the token. Default 3600
+        """
+        r_client.hset(token, 'timestamp', datetime.datetime.now())
+        r_client.hset(token, 'client_id', client_id)
+        r_client.expire(token, timeout)
+        if user:
+            r_client.hset(token, 'user', user)
+            # Check if client has access limit key, and if not, create it
+            limit_key = '%s_%s_daily_limit' % (client_id, user)
+            limiter = r_client.get(limit_key)
+            if limiter is None:
+                # Set limit to 5,000 requests per day
+                r_client.setex(limit_key, 5000, 86400)
+
+    def validate_client(self, client_id, client_secret):
+        """Make sure client exists, then set the token and send it
+
+        Parameters
+        ----------
+        client_id : str
+            The client making the request
+        client_secret : str
+            The secret key for the client
+        """
+        if login_client(client_id, client_secret):
+            token = self.generate_access_token()
+            self.write({'access_token': token,
+                        'token_type': 'Bearer',
+                        'expires_in': '3600'})
+            self.finish()
+            self.set_token(token, client_id)
+        else:
+            self.oauth_error('Oauth2 error: invalid client information',
+                             'invalid_client')
+
+    def validate_resource_owner(self, username, password, client_id):
+        """Make sure user and client exist, then set the token and send it
+
+        Parameters
+        ----------
+        username : str
+            The username to validate
+        password : str
+            The password for the username
+        client_id : str
+            The client making the request
+        """
+        try:
+            qdb.user.User.login(username, password)
+        except (IncorrectEmailError, IncorrectPasswordError,
+                UnverifiedEmailError):
+            self.oauth_error('Oauth2 error: invalid user information',
+                             'invalid_client')
+            return
+
+        if login_client(client_id):
+            token = self.generate_access_token()
+            self.write({'access_token': token,
+                        'token_type': 'Bearer',
+                        'expires_in': '3600'})
+            self.finish()
+            self.set_token(token, client_id, user=username)
+        else:
+            self.oauth_error('Oauth2 error: invalid client information',
+                             'invalid_client')
+
+    def post(self):
```

Comment:
Just to double check on how oauth2 works. You get a new token by posting? It looks counter-intuitive.